src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -346,24 +346,21 @@ internal sealed class CompletionSource : IAsyncExpandingCompletionSource
                 AsyncCompletionLogger.LogSourceGetContextTicksDataPoint((int)totalStopWatch.Elapsed.TotalMilliseconds, isCanceled: cancellationToken.IsCancellationRequested);
             }
 
-            static VSCompletionContext CombineCompletionContext(VSCompletionContext context1, VSCompletionContext context2)
+            static VSCompletionContext CombineCompletionContext(IAsyncCompletionSession session, VSCompletionContext context1, VSCompletionContext context2)
             {
-                if (context1.Items.IsEmpty && context1.SuggestionItemOptions is null)
+                if (context1.ItemList.IsEmpty && context1.SuggestionItemOptions is null)
                     return context2;
 
-                if (context2.Items.IsEmpty && context2.SuggestionItemOptions is null)
+                if (context2.ItemList.IsEmpty && context2.SuggestionItemOptions is null)
                     return context1;
 
-                using var _ = ArrayBuilder<VSCompletionItem>.GetInstance(context1.Items.Length + context2.Items.Length, out var itemsBuilder);
-                itemsBuilder.AddRange(context1.Items);
-                itemsBuilder.AddRange(context2.Items);
-
+                var completionList = session.CreateCompletionList(context1.ItemList.Concat(context2.ItemList));
```

Comment:
which .Concat is this?  is it a single alloc that just streams over both lists?  or is it an O(N) alloc that copies everything into a new list?

Reply:
I think this is the one streams over both lists w/o copying (`System.Linq.Enumerable.Concat`)